solr/core/src/java/org/apache/solr/cluster/placement/PropertyKeyFactory.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.cluster.placement;
+
+/**
+ * Factory used by the plugin to create property keys to request property values from Solr.<p>
+ *
+ * Building of a {@link PropertyKey} requires specifying the target (context) from which the value of that key should be
+ * obtained. This is done by specifying the appropriate {@link PropertyValueSource}.<br>
+ * For clarity, when only a single type of target is acceptable, the corresponding subtype of {@link PropertyValueSource} is used instead
+ * (for example {@link Node}).
+ */
+public interface PropertyKeyFactory {
+  /**
+   * Returns a property key to request the number of cores on a {@link Node}.
+   */
+  PropertyKey createCoreCountKey(Node node);
+
+  /**
+   * Returns a property key to request disk related info on a {@link Node}.
+   */
+  PropertyKey createDiskInfoKey(Node node);
+
+  /**
+   * Returns a property key to request the value of a system property on a {@link Node}.
+   * @param systemPropertyName the name of the system property to retrieve.
+   */
+  PropertyKey createSystemPropertyKey(Node node, String systemPropertyName);
+
+  /**
+   * Returns a property key to request the value of a metric.<p>
+   *
+   * Not all metrics make sense everywhere, but metrics can be applied to different objects. For example
+   * <code>SEARCHER.searcher.indexCommitSize</code> would make sense for a given replica of a given shard of a given collection,
+   * and possibly in other contexts.<p>
+   *
+   * @param metricSource The registry of the metric. For example a specific {@link Replica}.
+   * @param metricName for example <code>SEARCHER.searcher.indexCommitSize</code>.
+   */
+  PropertyKey createMetricKey(PropertyValueSource metricSource, String metricName);
```

Comment:
We have two sets of "global" metrics for a Solr instance: JVM and node-level. I understand we could use Node to select node-level metrics (from `solr.node` registry) but how to select `solr.jvm` metrics? Perhaps we need a `JVMPropertyValueSource`

Reply:
So these would be metrics that live on a node but that are accessed differently or with a different name? If we were able to distinguish by some other mean, would Node be an appropriate PropertyValueSource? Another type of PropertyValueSource can be introduced, but then would have to point to a specific JVM...
Can you point me to examples of these two metrics in 8x or trunk?

Reply:
`SolrDispatchFilter.setupJvmMetrics` initializes per-JVM metrics. They appear in a separate `solr.jvm` registry, which is different from `solr.node`.

In 99% cases (or practically always in production) Solr node maps 1:1 to a JVM instance. In some cases (most notably tests) there can be multiple Solr nodes running in a single JVM, so it's a N:1 - but never the other way around because it wouldn't make sense. So in some rare cases we will have multiple `solr.node` registries in one JVM (reachable via different API endpoints), but always a single `solr.jvm` registry (also reachable via different endpoints).

Reply:
.. and yes, the metrics in `solr.jvm` are a totally different set of metrics from those in `solr.node`.

Reply:
Maybe for metrics there's a need to add a registry name in addition to the `PropertyValueSource`? That would likely solve this issue.
My thinking was that we might have metrics on other things than a `Node` or the JVM where a `Node` runs. Like metrics on a collection...
If this is not the case, then I guess metrics should have a value source of `Node`, and then a registry name (that might better be expressed as enum if there are only two).

Reply:
Is `solr.jetty` of any use or `solr.node` and `solr.jvm` are sufficient?

I see `SolrInfoBean.Group` entries other than `node`, `jvm`, `jetty` are:
- `collection`, `shard`, `cluster`: these can be inferred from `PropertyValueSource` type,
- `core`: likely not needed for placement computation,
- `overseer`: doesn't seem to be used

Reply:
One node usually hosts many replicas. Each of these replicas has a unique registry name, in the form of `solr.core.<replicaName>`, so we could build PropertyKey from Replica because all components of the full metrics name are known.

This is not the case with `node`, `jvm` and `jetty` - I think we need to explicitly specify the registry name in these cases.